File: ielts-vocab-master/PHP/auth/reset_password.php
<?php
header('Content-Type: application/json');
require __DIR__ . '/db.php';

$body = json_decode(file_get_contents('php://input'), true);
$token = isset($body['token']) ? trim($body['token']) : '';
$password = $body['password'] ?? '';

if ($token === '') {
    http_response_code(400);
    echo json_encode(['error' => 'Missing reset token.']);
    exit;
}
if (strlen($password) < 8) {
    http_response_code(400);
    echo json_encode(['error' => 'Password must be at least 8 characters.']);
    exit;
}

$pdo = get_db();
$tokenHash = hash('sha256', $token);
$stmt = $pdo->prepare("SELECT id FROM users WHERE reset_token = ? AND reset_token_expires > datetime('now')");
$stmt->execute([$tokenHash]);
$user = $stmt->fetch(PDO::FETCH_ASSOC);

if (!$user) {
    http_response_code(400);
    echo json_encode(['error' => 'This reset link is invalid or has expired.']);
    exit;
}

$hash = password_hash($password, PASSWORD_DEFAULT);
$pdo->prepare('UPDATE users SET password_hash = ?, reset_token = NULL, reset_token_expires = NULL WHERE id = ?')
    ->execute([$hash, $user['id']]);

echo json_encode(['ok' => true]);

File: VocabLarry/Python/Flask/auth/reset_password.py
import hashlib

from flask import Blueprint, jsonify, request
from werkzeug.security import generate_password_hash

from db import get_db

reset_password_bp = Blueprint('reset_password', __name__)


@reset_password_bp.route('/reset_password', methods=['POST'])
def reset_password():
    body     = request.get_json(silent=True) or {}
    token    = body.get('token', '').strip()
    password = body.get('password', '')

    if not token:
        return jsonify({'error': 'Missing reset token.'}), 400
    if len(password) < 8:
        return jsonify({'error': 'Password must be at least 8 characters.'}), 400

    token_hash = hashlib.sha256(token.encode()).hexdigest()
    conn = get_db()
    user = conn.execute(
        "SELECT id FROM users WHERE reset_token = ? AND reset_token_expires > datetime('now')",
        (token_hash,),
    ).fetchone()

    if not user:
        return jsonify({'error': 'This reset link is invalid or has expired.'}), 400

    conn.execute(
        'UPDATE users SET password_hash = ?, reset_token = NULL, reset_token_expires = NULL WHERE id = ?',
        (generate_password_hash(password), user['id']),
    )
    conn.commit()
    return jsonify({'ok': True})
